Better bootstrap-aware Imp tests

teodanciu opened the pull request:
This PR is enabling conway Imp tests  to run with both protocol version 9 and 10.

Instead of having tests grouped in functions with respect to whether they should be run for both versions, each test is individually specifying whether it should run only for version 10 , in case there are operations involved that are disallowed in version 9. 

Where I felt it made sense, i have ran them for both versions, checking that the expected `DisallowedProposalDuringBootstrap` / `DisallowedVotesDuringBootstrap` predicate failure is returned - I chose to this where i thought that the test is more about submission of proposals, rather than the logic involving disallowed proposals. 


The PR contains a few other improvements - extracting some commonly used logic to functions, moving a few tests  where it seems like they would make more sense, removing some that didn't make sense to me. 


Because of this moving around of code, I preferred to keep the commits like this for the moment -  in order to make reviewing easier.  To avoid all this noise in master,  it would be better to squash at least all the commits that are enabling ImpTests for both versions - there is no point to have one commit per file in the end. 

Closes  https://github.com/IntersectMBO/cardano-ledger/issues/4703

# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Extract creating proposal from GovAction to `mkProposal`

teodanciu committed:
Introduce `submitBootstrapAwareFailingProposal`

teodanciu committed:
GovSpec - constitutionSpec

teodanciu committed:
GovSpec - proposalsWithVotingSpec, proposalsSpec

teodanciu committed:
Move test from `GovSpec` to more appropriate `EpochSpec`

since it is checking various states over several epochs

teodanciu committed:
Remove superfluous `submitConstitutionGovAction` function

and replace usages with similar `submitConstitution`

teodanciu committed:
GovSpec - unknownCostModelsSpec

teodanciu committed:
GovSpec - votingSpec

teodanciu committed:
GovSpec - policySpec, predicateFailureSpec, networkIdSpec

teodanciu committed:
GovSpec - withdrawalsSpec

teodanciu committed:
Make GovSpec tests pass for both Conway protocol versions

teodanciu committed:
Make UtxosSpec tests pass for both Conway protocol versions

teodanciu committed:
Add `mkUpdateCommitteeProposal` and `mkParameterChangeProposal`

and use them in their respective submit-function, for more flexible
usage

teodanciu committed:
Move tests from `GovCertSpec` to more appropriate `EnactSpec`

since they are testing the behavior of the proposal across eras based on
votes

teodanciu committed:
Make GovCertSpec tests pass for both Conway protocol versions

teodanciu committed:
Make EnactSpec tests pass for both Conway protocol versions

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -1771,3 +1803,60 @@ logConwayTxBalance tx = do
+  NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)) ->
+  SubmitFailureExpectation era
+pattern FailBoth {bootstrapFailures, postBootstrapFailures} =
+  FailBootstrapAndPostBootstrap bootstrapFailures postBootstrapFailures
```
What's the point of this type synonym? It is identical to `FailBootstrapAndPostBootstrap`

teodanciu replied:
In order to be able to give names  for `bootstrapFailures` and `postBootstrapFailures`. 
If i do this:
```
data SubmitFailureExpectation era
  = FailBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailPostBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailBootstrapAndPostBootstrap
      { bootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
      , postBootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
     } 
```
I get a compiler warning : "    Use of partial record field selector: ‘bootstrapFailures’" 

I guess without names it's pretty clear which is which, but I thought since they have the same type, it would be better to be able to pass with names the values.

lehins replied:
Using record syntax for sum types is indeed not a good idea, hence the compiler warning. With pattern synonym you have the same issue, you just manage to shut the compiler up.

The correct way would be to create an extra type (less overhead than pattern synonym too):

```
data SubmitFailureExpectation era
  = FailBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailPostBootstrap (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
  | FailBootstrapAndPostBootstrap (FailBoth era)

data FailBoth era = FailBoth
      { bootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
      , postBootstrapFailures ::  (NE.NonEmpty (PredicateFailure (EraRule "LEDGER" era)))
     } 
```

teodanciu replied:
Interesting!  Thank you

teodanciu committed:
Remove RatifySpec test with unclear purpose

The test seems to check that NoConfidence and CommitteUpdate govActions
are enacted during bootstrap period, but the test cannot run during
bootstrap, because these actions are not allowed.

teodanciu committed:
Make RatifySpec tests pass for both Conway protocol versions

teodanciu committed:
Make `EnactSpec` tests pass for both Conway protocol versions

teodanciu committed:
Run all Conway Imp tests for both version 9 and 10

teodanciu committed:
Fix CertsSpec for post-bootstrap

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -1771,3 +1792,56 @@ logConwayTxBalance tx = do
+submitBootstrapAwareFailingProposal proposal =
+  submitBootstrapAware
+    (SJust <$> submitProposal proposal)
+    (submitFailingProposal proposal >=> const (pure SNothing))
```
This can be a regular `Maybe`, instead of `StrictMaybe`. 
`const (pure Nothing)` can be turned into `Nothing <$`

Also we could a add a helper function that discards output, since that is the most common use case:
```suggestion
submitBootstrapAwareFailingProposal ::
  ConwayEraImp era =>
  ProposalProcedure era ->
  SubmitFailureExpectation era ->
  ImpTestM era (Maybe (GovActionId (EraCrypto era)))
submitBootstrapAwareFailingProposal proposal =
  submitBootstrapAware
    (Just <$> submitProposal proposal)
    ((Nothing <$) . submitFailingProposal proposal)

submitBootstrapAwareFailingProposal_ ::
  ConwayEraImp era =>
  ProposalProcedure era ->
  SubmitFailureExpectation era ->
  ImpTestM era ()
submitBootstrapAwareFailingProposal_ = void . submitBootstrapAwareFailingProposal
```

lehins commented on eras/conway/impl/CHANGELOG.md:
```diff
@@ -6,6 +6,11 @@
 ### `testlib`
 
+* Add `mkUpdateCommitteeProposal`
+* Add `SubmitFailureExpectation`, `submitBootstrapAwareFailingVote`, `submitBootstrapAwareFailingProposal`
```
```suggestion
* Add `SubmitFailureExpectation`, `FailBoth`, `submitBootstrapAwareFailingVote`, `submitBootstrapAwareFailingProposal`
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs:
```diff
@@ -85,12 +71,11 @@ unknownCostModelsSpec =
       gai <-
         submitParameterChange SNothing $
           emptyPParamsUpdate
             & ppuCostModelsL .~ SJust newCostModels
```
Instead of mocking with threshold I'd recommend only voting whenever we out of bootstrap:
```suggestion
      (drepC, _, _) <- setupSingleDRep 1_000_000
      gai <-
        submitParameterChange SNothing $
          emptyPParamsUpdate
            & ppuCostModelsL .~ SJust newCostModels
      whenPostBootstrap $ submitYesVote_ (DRepVoter drepC) gai
```

It is ok to register a DRep in the bootstrap phase

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/GovSpec.hs:
```diff
@@ -1190,81 +1067,77 @@ withdrawalsSpec =
+            proposal <- mkProposal wdrls
+            void $
+              submitBootstrapAwareFailingProposal proposal $
+                FailBootstrap [disallowedProposalFailure proposal]
```
With all the cleanup and deduplication there is no longer a need for `let in` usage, since there is only one `wdrls`.
```suggestion
      proposal <- mkProposal wdrls
      submitBootstrapAwareFailingProposal_ proposal $
        FailBootstrap [disallowedProposalFailure proposal]
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxosSpec.hs:
```diff
@@ -469,54 +453,41 @@ govPolicySpec ::
 govPolicySpec = do
   describe "Gov policy scripts" $ do
-    it "failing native script govPolicy" $ do
+    -- These tests rely on the script in the constitution, but we can only change the constitution after bootstrap.
```
This is only a problem because we don't have a guardrail script in the genesis file, which we should change. I've created a ticket about it #4726

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
@teodanciu This test is unclear because it evolved from a more clear test into something that expected to be tested in bootstrap and after bootstrap. This is a testament to the fact that this PR was very much needed.

If I remember correctly this test(s) where suppose to check SPOs voting on 

Instead of removing tests it would be better to adjust them to make them work in the new re-organized imp spec. I'd suggest one of them to run in the post-bootsrap only, while the other switching to using a different action, like PParamUpdate.

We don't have to do it in this PR, I'll create a separate ticket for it: https://github.com/IntersectMBO/cardano-ledger/issues/4727

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp.hs:
```diff
@@ -97,28 +97,20 @@ spec ::
+          describe "UTXO" $ Utxo.spec @era
+          describe "UTXOS" $ Utxos.spec @era
+   in
+    forM_ [natVersion @9, natVersion @10] conwayImpSpec
```
This will ensure that we are running for all protocol versions, just in case we have another intra-era hard fork
```suggestion
    forM_ [eraProtVerLow @Conway .. eraProtVerHigh @Conway] conwayImpSpec
```

lehins reviewed (approved):
Awesome cleanup and general improvement of imp specs!
Thank you!
